File: Solutions/98-IsValidBinarySearchTree/IsValidBinarySearchTree.kt
class Solution {
    fun isValidBST(root: TreeNode?): Boolean {
        return dfs(root, null, null)   
    }
    
    fun dfs(node: TreeNode?, min: Int?, max: Int?): Boolean {
        if(node == null) return true;

        if((min != null && node.`val` <= min) || (max != null && node.`val` >= max)) return false;

        val isLeftValid = dfs(node.left, min, node.`val`)
        val isRightValid = dfs(node.right, node.`val`, max)
        return isLeftValid && isRightValid
    }
}

File: Solutions/All/98-IsValidBinarySearchTree/IsValidBinarySearchTree.ts
import { TreeNode } from "../../../Types";

export function isValidBST(root: TreeNode | null): boolean {
  return dfs(root, -Infinity, Infinity);
}

function dfs(node: TreeNode | null, min: number, max: number): boolean {
  if (!node) return true;

  if (node.val <= min || node.val >= max) return false;

  const isLeftValid = dfs(node.left, min, node.val);
  const isRightValid = dfs(node.right, node.val, max);
  return isLeftValid && isRightValid;
}
